draft-ietf-privacypass-architecture.md

```diff
@@ -174,14 +132,69 @@ The following terms are used throughout this document.
   attested to by the Attester.
 - Attester: An entity that attests to properties of Client for the
   purposes of token issuance.
+- Redemption context: The interactions and set of information shared
+between the Client and Origin.
+- Issuance context: The interactions and set of information shared
+between the Client, Attester, and Issuer.
 
 # Architecture
 
 The Privacy Pass architecture consists of four logical entities --
 Client, Origin, Issuer, and Attester -- that work in concert as
-shown in {{introduction}} for token issuance and redemption. This
-section describes the purpose of token issuance and redemption
-and the requirements therein on the relevant participants.
+for token issuance and redemption. This section describes the purpose
+of token issuance and redemption and the requirements on the relevant
+participants.
+
+The typical interaction flow for Privacy Pass tokens uses the
+following steps:
+
+1. A Client interacts with an Origin by sending an HTTP request.
+The Origin sends an HTTP response that contains a token challenge
+that indicates a specific Issuer to use.
+Note that the request might be made as part of accessing a
+resource normally, or with the specific intent of triggering a token
+challenge.
+
+2. If the Client already has a token available that satisfies the token
+challenge, it can skip to step 5 and redeem its token. Otherwise, it
+performs attestation with the Attester and creates a Token Request to send
+to the Issuer (generally via the Attester).
+
+3. The Attester performs attestation checks on the Client. These checks
+could be proof of solving a CAPTCHA, device trust, hardware attestation,
+etc (see {{attester}}). If these checks pass, it sends the Token Request
+to the Issuer. The Attester and Issuer might be functions on the same server,
+depending on the deployment model (see {{deployment}}). If the checks fail,
+the Client receives an error.
+
+4. The Issuer generates a Token Response based on the Token Request, which
+is returned to the Client. Upon receiving the Token Response, the Client
+generates a token by cryptographically transforming the content of the
+response so that the token can be validated using a per-Issuer key, but
+cannot be linked to the content of the Token Response.
+
+5. If the Client has a token, it includes it in a subsequent HTTP
+request to the Origin, as authorization. This token is sent only once.
+The Origin validates that the token was generated by the expected Issuer.
+
+~~~ aasvg
+    Origin           Client      Attester     Issuer
+                   +---------------------------------.
+     Request <-----+                                  |
+                   |                                  |
+ TokenChallenge --->                                  |
+                   | TokenRequest ---------->         |
+                   |             (attest)             |
```

Comment:
```suggestion
                   | <--- (attest) --->             |
                   | TokenRequest ---------->         |
```

I think it's important to make clear that attestation can happen before the token request and response, and not necessarily something in parallel. 

Also, we use different notation to show the process of attestation in the deployment diagrams ("----> Attest"). I'm wondering if there's a better way to draw this to make it clear that attestation is _not_ specified here, is an entirely separate and deployment specific protocol, etc. Maybe something like this?

~~~ aasvg
    Origin           Client      Attester     Issuer
     Request <-----+                                  
                                                       
 TokenChallenge --->                                  
                    +-----------------.
                    |    Attestation   |
                     `----------------'  
                      TokenRequest ---------->
~~~

Reply:
How about this as the full picture?
```
    Origin           Client         Attester     Issuer
    
     Request <------                                  
                                                       
 TokenChallenge --->   
                      <- (Attestation) ->                               
  
                      TokenRequest -------------->
                              <-------------- TokenResponse
 Request + Token <--
```

Reply:
This diagram has been confusing me since the very first time I looked at the draft. Since the process initiates with a client, can we please put the client on the leftmost column?

Reply:
The issue is that the client has an interaction with the origin on one hand, and with the attester/issuer on the other, and they are separated contexts.

@chris-wood thoughts?